ledger/acctupdates.go

```diff
@@ -288,12 +325,253 @@ func (au *accountUpdates) close() {
+		}
+
+	}
+}
+
+func (au *accountUpdates) LookupKeysByPrefix(round basics.Round, keyPrefix string, maxKeyNum uint64) ([]string, error) {
+	return au.lookupKeysByPrefix(round, keyPrefix, maxKeyNum, true /* take lock */)
+}
+
+func (au *accountUpdates) lookupKeysByPrefix(round basics.Round, keyPrefix string, maxKeyNum uint64, synchronized bool) (resultKeys []string, err error) {
+	var results map[string]bool
+	// keep track of the number of result key with value
+	var resultCount uint64
+
+	needUnlock := false
+	if synchronized {
+		au.accountsMu.RLock()
+		needUnlock = true
+	}
+	defer func() {
+		if needUnlock {
+			au.accountsMu.RUnlock()
+		}
+		// preparation of result happens in deferring function
+		// prepare result only when err != nil
+		if err == nil {
+			resultKeys = make([]string, 0, resultCount)
+			for resKey, present := range results {
+				if present {
+					resultKeys = append(resultKeys, resKey)
+				}
+			}
+		}
+	}()
+
+	// TODO: This loop and round handling is copied from other routines like
+	// lookupResource. I believe that it is overly cautious, as it always reruns
+	// the lookup if the DB round does not match the expected round. However, as
+	// long as the db round has not advanced too far (greater than `rnd`), I
+	// believe it would be valid to use. In the interest of minimizing changes,
+	// I'm not doing that now.
+
+	for {
+		currentDBRound := au.cachedDBRound
+		currentDeltaLen := len(au.deltas)
+		offset, _err := au.roundOffset(round)
+		if _err != nil {
+			err = _err
+			return
+		}
+
+		// reset `results` to be empty each iteration
+		// if db round does not match the round number returned from DB query, start over again
+		// NOTE: `results` is maintained as we walk backwards from the latest round, to DB
+		// IT IS NOT SIMPLY A SET STORING KEY NAMES!
+		// - if the boolean for the key is true: we consider the key is still valid in later round
+		// - otherwise, we consider that the key is deleted in later round, and we will not return it as part of result
+		// Thus: `resultCount` keeps track of how many VALID keys in the `results`
+		// DO NOT TRY `len(results)` TO SEE NUMBER OF VALID KEYS!
+		results = map[string]bool{}
+		resultCount = 0
+
+		for i := int(offset - 1); i >= 0; i-- {
+			for keyInRound, mv := range au.kvDeltas[i] {
+				if !strings.HasPrefix(keyInRound, keyPrefix) {
+					continue
+				}
+				// whether it is set or deleted in later round, if such modification exists in later round
+				// we just ignore the earlier insert
+				if _, ok := results[keyInRound]; ok {
+					continue
+				}
+				if mv.Data == nil {
+					results[keyInRound] = false
+				} else {
+					// set such key to be valid with value
+					results[keyInRound] = true
+					resultCount++
+					// check if the size of `results` reaches `maxKeyNum`
+					// if so just return the list of keys
+					if maxKeyNum > 0 && resultCount == maxKeyNum {
+						return
```

Comment:
returning here might only give the latest N boxes, not N of ordered boxes

Reply:
I don't think we make any promises about _which_ key names are returned if more than max key names exist.  I think the same thing is is true for getting assets or apps, though I might be mistake on this.  Since `algod` doesn't have a paging API for this call, it's not super helpful to know you got the first answers - there's no way to ask for the next answers.

Reply:
Resolving - We've decided to rework the API as is summarized in https://github.com/algorand/go-algorand/pull/4149#discussion_r1001083810.  Let's move discussion to https://github.com/algorand/go-algorand/pull/4149#discussion_r1001083810.